File: SCT2/trunk/test-plugins/org.yakindu.sct.simulation.core.sexec.test/test-gen/org/yakindu/sct/simulation/core/sexec/test/ShallowHistoryWithDeepEntryTest.java
/**
 * Copyright (c) 2014 committers of YAKINDU and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     committers of YAKINDU - initial API and implementation
 */
package org.yakindu.sct.simulation.core.sexec.test;
import org.eclipse.xtext.junit4.InjectWith;
import org.eclipse.xtext.junit4.XtextRunner;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.yakindu.sct.model.sexec.ExecutionFlow;
import org.yakindu.sct.model.sexec.interpreter.test.util.AbstractExecutionFlowTest;
import org.yakindu.sct.model.sexec.interpreter.test.util.SExecInjectionProvider;
import org.yakindu.sct.test.models.SCTUnitTestModels;
import com.google.inject.Inject;
import static org.junit.Assert.assertTrue;
/**
 *  Unit TestCase for ShallowHistoryWithDeepEntry
 */
@SuppressWarnings("all")
@RunWith(XtextRunner.class)
@InjectWith(SExecInjectionProvider.class)
public class ShallowHistoryWithDeepEntryTest extends AbstractExecutionFlowTest {
	@Before
	public void setup() throws Exception {
		ExecutionFlow flow = models
				.loadExecutionFlowFromResource("ShallowHistoryWithDeepEntry.sct");
		initInterpreter(flow);
	}
	@Test
	public void noDeepEntryWithinHistory() throws Exception {
		interpreter.enter();
		assertTrue(isActive("Y"));
		raiseEvent("toZ");
		interpreter.runCycle();
		assertTrue(isActive("A"));
		raiseEvent("toY");
		interpreter.runCycle();
		assertTrue(isActive("Y"));
		raiseEvent("toZ");
		interpreter.runCycle();
		assertTrue(isActive("A"));
	}
	@Test
	public void deepEntryWithinHistory() throws Exception {
		interpreter.enter();
		assertTrue(isActive("Y"));
		raiseEvent("toZ");
		interpreter.runCycle();
		assertTrue(isActive("A"));
		raiseEvent("toC");
		interpreter.runCycle();
		assertTrue(isActive("C"));
		assertTrue(isActive("B"));
		raiseEvent("toY");
		interpreter.runCycle();
		assertTrue(isActive("Y"));
		raiseEvent("toZ");
		interpreter.runCycle();
		assertTrue(isActive("C"));
		assertTrue(isActive("B"));
	}
	@Test
	public void directDeepEntryIntoHistory() throws Exception {
		interpreter.enter();
		assertTrue(isActive("Y"));
		raiseEvent("toC");
		interpreter.runCycle();
		assertTrue(isActive("C"));
		assertTrue(isActive("B"));
		raiseEvent("toY");
		interpreter.runCycle();
		assertTrue(isActive("Y"));
		raiseEvent("toZ");
		interpreter.runCycle();
		assertTrue(isActive("C"));
		assertTrue(isActive("B"));
	}
}

File: test-plugins/org.yakindu.sct.generator.c.test/gtests/ShallowHistoryWithDeepEntry/ShallowHistoryWithDeepEntryTest.cc
/**
* Copyright (c) 2017 committers of YAKINDU and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v1.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v10.html
*
* Contributors:
*     committers of YAKINDU - initial API and implementation
*/
#include "gtest/gtest.h"
#include "ShallowHistoryWithDeepEntry.h"

static ShallowHistoryWithDeepEntry statechart;

class StatemachineTest : public ::testing::Test{
	protected:
	virtual void SetUp() {
		shallowHistoryWithDeepEntry_init(&statechart);
	}
};


TEST_F(StatemachineTest, noDeepEntryWithinHistory) {					
	shallowHistoryWithDeepEntry_enter(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Y));
	shallowHistoryWithDeepEntryIface_raise_toZ(&statechart);
	shallowHistoryWithDeepEntry_runCycle(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_A));
	shallowHistoryWithDeepEntryIface_raise_toY(&statechart);
	shallowHistoryWithDeepEntry_runCycle(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Y));
	shallowHistoryWithDeepEntryIface_raise_toZ(&statechart);
	shallowHistoryWithDeepEntry_runCycle(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_A));
}
TEST_F(StatemachineTest, deepEntryWithinHistory) {					
	shallowHistoryWithDeepEntry_enter(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Y));
	shallowHistoryWithDeepEntryIface_raise_toZ(&statechart);
	shallowHistoryWithDeepEntry_runCycle(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_A));
	shallowHistoryWithDeepEntryIface_raise_toC(&statechart);
	shallowHistoryWithDeepEntry_runCycle(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_B__region0_C));
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_B));
	shallowHistoryWithDeepEntryIface_raise_toY(&statechart);
	shallowHistoryWithDeepEntry_runCycle(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Y));
	shallowHistoryWithDeepEntryIface_raise_toZ(&statechart);
	shallowHistoryWithDeepEntry_runCycle(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_B__region0_C));
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_B));
}
TEST_F(StatemachineTest, directDeepEntryIntoHistory) {					
	shallowHistoryWithDeepEntry_enter(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Y));
	shallowHistoryWithDeepEntryIface_raise_toC(&statechart);
	shallowHistoryWithDeepEntry_runCycle(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_B__region0_C));
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_B));
	shallowHistoryWithDeepEntryIface_raise_toY(&statechart);
	shallowHistoryWithDeepEntry_runCycle(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Y));
	shallowHistoryWithDeepEntryIface_raise_toZ(&statechart);
	shallowHistoryWithDeepEntry_runCycle(&statechart);
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_B__region0_C));
	EXPECT_TRUE(shallowHistoryWithDeepEntry_isStateActive(&statechart, ShallowHistoryWithDeepEntry_main_region_Z__region0_B));
}


